source/packages/libraries/core/version/scripts/compile.ts

```diff
@@ -0,0 +1,48 @@
+import fs from 'fs';
+import readline from 'readline';
+import { Project } from 'ts-morph';
+
+async function extractVersion(): Promise<string> {
+    const versionPoliciesReadStream = fs.createReadStream(
+        '../../../../common/config/rush/version-policies.json'
+    );
+    const rl = readline.createInterface({
+        input: versionPoliciesReadStream,
+        crlfDelay: Infinity,
+    });
+
+    let previousLine: string = '';
+    let versionLine: string | undefined = undefined;
+    for await (const line of rl) {
+        if (previousLine.includes('"policyName": "external",')) {
+            versionLine = line;
+            break;
+        }
+
+        previousLine = line;
+    }
+
+    // https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+    const versionRegex =
+        /(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?/;
+    const version = versionLine!.match(versionRegex);
+
+    return version![0];
```

Comment:
This should error before here if the regex returns no matches.

Reply:
The question is when is it possible to get no match? What are we guardrailing here? I'm using non-null assertion as described in [TypeScript doc](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-)

> Just like other type assertions, this doesn’t change the runtime behavior of your code, so it’s important to only use ! when you know that the value can’t be null or undefined.

Reply:
For known, obvious errors, we should provide useful error logs or protect against them.

Yes, it's unlikely, but the previous versionLine can error if we change things like the policyName or the order of the definition in the version-policies file. This line will additionally fail if that regex doesn't match the version definition (which is hard to reason about quickly). We can head both of these off by either having a sane defaults for missing values or catching them with a nice error that says "version-policies.json does not contain version for policyName: external" and "Version for policyName: external is (insert found versionLine). No version matching regex found."

As it describes, this doesn't change the runtime behavior, so if for any reason either of these is missed, you're going to get a "cannot access property of null or undefined" error which isn't super useful.